department-of-veterans-affairs/caseflow
# MattT/Fix EMO Assigned Column Task Filtration

Re-Resolves [CASEFLOW-4600](https://vajira.max.gov/browse/CASEFLOW-4600)


### Description
Fix the task filtering for the EMO Assigned queue tab to display the proper tasks (PreDocketTasks with status `assigned` and EducationAssessDocumentationTasks with statuses `assigned` or `in_progress`.

### Acceptance Criteria
- [ ] Determination of the behavior and triggers of the Assigned tab
- [ ] When cases are assigned to one of the three (3) RPO organizations, they appear on the Assigned tab
- [ ] If the RPO organization sends it to board intake, the appeal remains in EMO’s Assigned tab (it is also moved to the Completed tab of the RPO queue)
- [ ] If the RPO organization sends the appeal back to the EMO office, it is removed from the Assigned tab goes to the Unassigned tab
- [ ] If the EMO organization sends an appeal to BVA Intake, regardless of the reason selected (not an education issue or appeal ready for review), the appeal shows on the EMO's "Assigned" tab.
- [ ] Education-related appeals are able to be docketed with only the `docket_vha_appeals` FeatureToggle enabled (and not `docket_edu_appeals`).

### EMO Assigned Tab Testing Plan
1. Ensure that you have the `edu_predocket_appeals` and `docket_vha_appeals` feature toggles enabled:

```ruby
FeatureToggle.enable!(:edu_predocket_appeals)
FeatureToggle.enable!(:docket_vha_appeals)
```

2. Create an appeal with an `EducationDocumentSearchTask`.
```ruby
emo_task = FactoryBot.create(:education_document_search_task, :assigned)
appeal = emo_task.appeal
```

3. Log into Caseflow as an EMO user. Go to the EMO's [unassigned queue tab](http://localhost:3000/organizations/edu-emo?tab=education_unassigned) and make sure that the appeal appears there.

4. Either use the Caseflow UI or run the following commands to simulate sending the appeal back to BVA Intake. Afterwards, check that the appeal is listed on the [EMO's assigned queue tab](http://localhost:3000/organizations/edu-emo?tab=education_assigned) as a `Pre Docket` task assigned to `BVA Intake`.

```ruby
emo_task.update!(status: Constants.TASK_STATUSES.completed)
emo_task.parent.update!(status: Constants.TASK_STATUSES.assigned)
```

5. Reset the appeal, and this time simulate the scenario where the EMO has sent an appeal to an RPO. Afterwards, ensure that the appeal appears on the [EMO's assigned queue tab](http://localhost:3000/organizations/edu-emo?tab=education_assigned) as an `Assess Documentation` task assigned to one of the RPOs.

```ruby
emo_task.update!(status: Constants.TASK_STATUSES.on_hold)
emo_task.parent.update!(status: Constants.TASK_STATUSES.on_hold)

rpo_task = EducationAssessDocumentationTask.create!( parent: emo_task, appeal: appeal, assigned_at: Time.zone.now, assigned_to: EduRegionalProcessingOffice.first)
```

6. Set the `rpo_task` to `in_progress` and confirm that the appeal is still listed on the [EMO's assigned queue tab](http://localhost:3000/organizations/edu-emo?tab=education_assigned).

```ruby
rpo_task.update!(status: Constants.TASK_STATUSES.in_progress)
```

7. Now, simulate an RPO sending an appeal back to the EMO. After staging the scenario, check that the appeal is now on the [EMO's unassigned queue tab](http://localhost:3000/organizations/edu-emo?tab=education_unassigned).

```ruby
rpo_task.update!(status: Constants.TASK_STATUSES.completed)
emo_task.update!(status: Constants.TASK_STATUSES.assigned)
```

8. Next, send the appeal back to an RPO. The appeal should once again appear in the [EMO's assigned queue tab](http://localhost:3000/organizations/edu-emo?tab=education_assigned). Ensure it only appears once, because now there will be two child `EducationAssessDocumentationTasks`. IT will be listed as an `Assess Documenation` task assigned to an RPO.

```ruby
emo_task.update!(status: Constants.TASK_STATUSES.on_hold)
rpo_task2 = EducationAssessDocumentationTask.create!( parent: emo_task, appeal: appeal, assigned_at: Time.zone.now, assigned_to: EduRegionalProcessingOffice.last)
```

9. Simulate an RPO sending the appeal straight to BVA Intake. The appeal should appear in the [EMO's assigned queue tab](http://localhost:3000/organizations/edu-emo?tab=education_assigned) as a `Pre Docket` task assigned to `BVA Intake`.

```ruby
rpo_task2.update!(status: Constants.TASK_STATUSES.completed)
emo_task.update!(status: Constants.TASK_STATUSES.completed)
emo_task.parent.update!(status: Constants.TASK_STATUSES.assigned)
```

10. Set the appeal to appear as though BVA Intake has docketed it. It should no longer appear in the EMO's unassigned or assigned tabs, and rather be shown in its completed tab (not yet available).

```ruby
emo_task.parent.update!(status: Constants.TASK_STATUSES.completed)
```

11. Finally, roll back the docketing and test the scenario where BVA Intake returns an appeal to the EMO. It should appear in the [EMO's unassigned queue tab](http://localhost:3000/organizations/edu-emo?tab=education_unassigned):

```ruby
emo_task.parent.update!(status: Constants.TASK_STATUSES.on_hold)
emo_task2 = EducationDocumentSearchTask.create!( parent: emo_task.parent, appeal: appeal, assigned_at: Time.zone.now, assigned_to: EducationEmo)
```

### `docket_edu_appeals` Feature Toggle Removal Testing Plan

1. Disable the `docket_edu_appeals` feature toggle:

```ruby
FeatureToggle.disable!(:docket_edu_appeals)
```

2. Enable the `docket_vha_appeals` feature toggle:

```ruby
FeatureToggle.enable!(:docket_vha_appeals)
```

3. Create another pre-docketed education appeal, and set it to having been sent to BVA Intake for review:

```ruby
emo_task = FactoryBot.create(:education_document_search_task, :assigned)
emo_task.update!(status: Constants.TASK_STATUSES.completed)
emo_task.parent.update!(status: Constants.TASK_STATUSES.assigned)
edu_appeal2 = emo_task.appeal
```
You will revisit this appeal in step 5.

4. Disable the `docket_vha_appeals` feature toggles:

```ruby
FeatureToggle.disable!(:vha_predocket_appeals)
FeatureToggle.disable!(:docket_vha_appeals)
```                              

5. Get the UUID for `edu_appeal2` from step 3.

```ruby
edu_appeal2.uuid

# http://localhost:3000/queue/appeals/<uuid>
```

Navigate to its queue page and confirm, as a BVA Intake user without the `docket_vha_appeals` feature toggle enabled, that you are unable to docket the appeal from the actions dropdown.



### User Facing Changes

 There is no before image available for this change as it is a brand new UI addition.


![localhost_3000_organizations_edu-emo (1)](https://user-images.githubusercontent.com/99351305/165395458-6d8b202f-3875-4d42-ba54-943780e4070d.png)
